### PR TITLE
logind: add support for hidraw devices (via eBPF-based revoke)

### PR DESCRIPTION
**Pull request by whot**
*This is the replacement of #23140, using eBPF instead of a new ioctl to revoke access to hidraw devices.*

Add support for opening `/dev/hidraw` devices via logind's TakeDevice().
Same semantics as our support for evdev devices, but it requires eBPF support to revoke the fd.

# Requirements

- kernel: the WIP kernel patches required for this to work are [here](https://gitlab.freedesktop.org/bentiss/hid/-/tree/wip/hidraw-revoke-tests?ref_type=heads).
- selinux: https://github.com/fedora-selinux/selinux-policy/pull/1920 is required for SELinux to be happy.
- libinput: The easiest (only) way to test this is with [this libinput branch](https://gitlab.freedesktop.org/whot/libinput/-/commits/wip/hidraw-access) which requests all `/dev/hidraw` devices to be opened and then leaks the fd (i.e. keeps it open). On logout the devices will be revoked, staring hard at the journal will show that the revoke bit works.

cc @bentiss, @GeorgesStavracas, @swick, @hadess, @jadahl, @garnacho

### CODE REVIEW

File: src/login/logind-session-device.c [hunk 1]
```diff
@@ -275,21 +275,26 @@ static int session_device_verify(SessionDevice *sd) {
...
-                /* Prevent opening unsupported devices. Especially devices of
-                 * subsystem "input" must be opened via the evdev node as
-                 * we require EVIOCREVOKE. */
+        case DEVICE_TYPE_DRM:
+                break;
+
+        case  DEVICE_TYPE_UNKNOWN:
+        default:
                 return -ENODEV;
+        }
```
Comment (poettering) [inline src/login/logind-session-device.c:408]: i cherry picked this commit directory into git main, since it looks trivial enough, makes general sense and your PR already passed CI with this.

File: src/login/bpf/logind-revoke.bpf.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+struct hidraw_revoke_syscall_args {
+       int fd;
+};
+
+SEC("syscall")
+int hidraw_revoke(struct hidraw_revoke_syscall_args *args)
+{
+       return hid_bpf_hidraw_revoke(args->fd);
+}
```
Comment (poettering) [inline src/login/bpf/logind-revoke.bpf.c:25]: even in bpf code we generally try to follow our coding style, i.e. `{` on same line as function name, but doesn't really matter

File: src/login/logind-session-device.c [hunk 3]
```diff
@@ -103,6 +103,82 @@ static void sd_eviocrevoke(int fd) {
...
+                        );
+
+        assert(fd >= 0);
+
+        r = hidraw_bpf_attach(&obj);
+        if (r == 0) {
+                int bpf_fd = sym_bpf_program__fd(obj->progs.hidraw_revoke);
+                r = sym_bpf_prog_test_run_opts(bpf_fd, &tattrs);
+                if (r < 0)
+                        log_warning_errno(-r, "Revoking hidraw failed: %m");
```
Comment (poettering) [inline src/login/logind-session-device.c:172]: no need to add `-`. our log functions take abs() of the param anyway, hence you can pass either positive or negative doesn't atter

File: src/login/bpf/logind-revoke.bpf.c (hunk 2 above)
Comment (poettering) [reply src/login/bpf/logind-revoke.bpf.c:25]: btw, what's the rationale of exposing this via ebpf? i mean, the call is simply enough to just be an ioctl, no?

sounds like quite some heavy machinery for something that could just be an ioctl? can you explain the background of this?

Comment (bentiss) [reply src/login/bpf/logind-revoke.bpf.c:25]: > sounds like quite some heavy machinery for something that could just be an ioctl? can you explain the background of this?

Completely agree, and TBH I can not convince myself  it's the right way to go (which also explain why this wasn ´t sent to the LKML). So dumping my thoughts:

On the "why should we use eBPF" we have:
- eBPF kfuncs are considered as kABI, in the same way `EXPORT_GPL_SYMBOL` are. Which means that if we get the API wrong, we can always change it, it's kernel-to-kernel API
- eBPF kfuncs are validated by the eBPF verifier. So if we decide to change the API because of the previous point, we can guarantee that logind will be using the proper behavior
- we need eBPF for firewalling devices in the long run, so having the mechanics in place now kind of makes sense
- we can skip the fight for a generic "revoke" approach on the ioctl front, because it's going to be very hard to work.
- we could use a similar eBPF kfunc to revoke a fd owned by a different process. We can then add a "monitoring" eBPF program with logind that would also check for any other opened hidraw nodes, and include them in the revoke part when we need to do it.

On the "why using eBPF is pointless":
- the new ioctl is simple enough an known enough (but is uapi, so can not easily been changed)
- it's a heavy machinery to just get a single syscall
- the  "let's revoke other processes' fd" use case is not actually implemented, and might never be

So from the kernel side, I think the main advantage to using eBPF here is to not tie ourselves with a given API, meaning that we can extend/restrict it further without having to worry too much.

Opinions/thoughts?

Comment (poettering) [reply src/login/bpf/logind-revoke.bpf.c:25]: > * eBPF kfuncs are considered as kABI, in the same way `EXPORT_GPL_SYMBOL` are. Which means that if we get the API wrong, we can always change it, it's kernel-to-kernel API

Are you sure Linus really sees it that way? If we start using this symbol from userspace for things that are not purely tracing, are you sure you can just take it away again?

I mean, I can see that this is politically easier, I sympathize with that, but as I understand Linux once it is loadbearing for userspace it's not that easy to remove again.

Comment (poettering) [reply src/login/bpf/logind-revoke.bpf.c:25]: but frankly, I am fine with either. You have to fight the political fights on the kernel side. If this makes it simpler for you we are fine with the bpf approach. we are all in on bpf, no issues there.

Comment (bentiss) [reply src/login/bpf/logind-revoke.bpf.c:25]: > Are you sure Linus really sees it that way? If we start using this symbol from userspace for things that are not purely tracing, are you sure you can just take it away again?

I sneaked in last year kernel summit when this very same topic was discussed (and for that reason). His answer was definitely that eBPF is kernel-to-kernel, and that the only users of it are either big corporations running this on millions of machines, or systemd, in which case it's not a "normal" user, as they can know better on how to handle the situation.

It's even explicitly mentioned on the kernel docs:
https://docs.kernel.org/bpf/kfuncs.html#bpf-kfunc-lifecycle-expectations

> kfuncs provide a kernel <-> kernel API, and thus are not bound by any of the strict stability restrictions associated with kernel <-> user UAPIs. This means they can be thought of as similar to EXPORT_SYMBOL_GPL, and can therefore be modified or removed by a maintainer of the subsystem they're defined in when it's deemed necessary.
>
> Like any other change to the kernel, maintainers will not change or remove a kfunc without having a reasonable justification.

So basically, as I see it, if we agree for using a kfunc as a system call, logind should have a way to "test" if the kfunc is available, by attempting to load it once, and keep a boolean somewhere. This way, if that kfunc were to change, logind wouldn't break, the feature would not be available.

I think the easiest would be to attach the revoke BPF program only once, when the first a user of the hidraw fd appears, and only remove it when the last user of the last hidraw fd vanished. This should remove part of the heaviness by loading the bpf program once, and we can even try different versions of it, without doing such dance every time we need to revoke a hidraw fd.

Comment (whot) [reply src/login/bpf/logind-revoke.bpf.c:25]: Just chiming in with a summary of yesterday's chat with @bentiss:

`HIDIOCREVOKE` has a very specific use-case that we know (because we've done it with `EVIOCREVOKE`) and it's a tiny kernel patch and trivial to maintain.

BPF-based revoke can give us the same result with more effort and it has much greater flexibility. Alas, the use-cases we have for that greater flexibility are mostly theoretical, we don't know if we need them (things like revoking someone else's fd) and/or how we'd get to the point where we definitely need those. Other things we know we'll need BPF for (firewalling parts of the device) are orthogonal to revoke.

So right now the pendulum of motivation is swinging back to the ioctl :)

Comment (GeorgesStavracas) [reply src/login/bpf/logind-revoke.bpf.c:25]: @whot after making significant progress in the USB portal (and understanding the problem domain a bit more to believe I can provide feedback on this HID PR), I think we'll need a way to revoke device access from other apps.

Comment (whot) [reply src/login/bpf/logind-revoke.bpf.c:25]: @GeorgesStavracas - just to confirm, you need to be able to revoke an fd that you didn't open and pass along (like logind does)? Can you provide a few more details on the use-case here please?

Comment (swick) [reply src/login/bpf/logind-revoke.bpf.c:25]: We discussed this yesterday on matrix and `HIDIOCREVOKE` should be fine. What wasn't so clear to me however is if logind will only ever give us one file description, if it's possible to create another file description (maybe by re-opening the fd? But then the revoke mechanism from logind would not work anymore) and if it even ever makes sense to have multiple file descriptions to lease out concurrently to different apps/processes.

Comment (whot) [reply src/login/bpf/logind-revoke.bpf.c:25]: Right now logind only opens once (see [method_take_device](https://github.com/systemd/systemd/blob/main/src/login/logind-session-dbus.c#L482)). That makes sense for evdev (well, mice/keyboards) but not so much for hidraw so I'd say we need to extend the logind API to be able to re-open a device so they can be individually revoked.

**Comment by whot** [comment]
Ok, so long story short I am now closing this MR and we're switching back to the ioctl approach as in #23140. This decision was made by @bentiss and I a few months ago (shortly after the above comments actually) but we've been ETIME to do something about it until now.

In summary: there's a specific use-case that we want to fix that and that is to provide revoke for hidraw devices via logind. Long-term (just like evdev devices) this is the best approach at providing access to HID devices to userspace clients though it obv. needs more infrastructure in place to e.g. forward the fd via Wayland to a real client.

The BPF approach works but has a few drawbacks and is vastly more complicated. There are other BPF revoke-like use-cases but none of them need be in logind and they don't fully address what we need to do either.


This PR is now superseded by #33970 (because I cannot re-open the original #23140).
